modules/pubxaiAnalyticsAdapter.js

```diff
@@ -1,208 +1,429 @@
+ * back to pubx.ai. The data may be removed from this cache, post send.
+ */
+export const sendCache = new Proxy(
+  {},
+  {
+    get: (target, name) => {
+      if (!target.hasOwnProperty(name)) {
+        target[name] = [];
+      }
+      return target[name];
+    },
+  }
+);
+
+/**
+ * auctionCache is a global cache object which stores all auction histories
+ * for the session. When getting a key from the auction cache, any
+ * information already known about the auction or associated data (floor
+ * data configured by prebid, browser data, user data etc) is added to
+ * the cache automatically.
+ */
+export const auctionCache = new Proxy(
+  {},
+  {
+    get: (target, name) => {
+      if (!target.hasOwnProperty(name)) {
+        target[name] = {
+          bids: [],
+          auctionDetail: {
+            refreshRank: Object.keys(target).length,
+            auctionId: name,
+          },
+          floorDetail: {},
+          pageDetail: {
+            host: getWindowLocation().host,
+            path: getWindowLocation().pathname,
+            search: getWindowLocation().search,
+          },
+          deviceDetail: {
+            platform: navigator.platform,
+            deviceType: getDeviceType(),
+            deviceOS: getOS(),
+            browser: getBrowser(),
+          },
+          userDetail: {
+            userIdTypes: Object.keys(getGlobal().getUserIds?.() || {}),
+          },
+          consentDetail: {
+            consentTypes: Object.keys(getGlobal().getConsentMetadata?.() || {}),
+          },
+          pmacDetail: JSON.parse(getStorage()?.getItem('pubx:pmac')) || {}, // {auction_1: {floor:0.23,maxBid:0.34,bidCount:3},auction_2:{floor:0.13,maxBid:0.14,bidCount:2}
+          initOptions: {
+            ...initOptions,
+            auctionId: name, // back-compat
+          },
+          sendAs: [],
+        };
+      }
+      return target[name];
+    },
+  }
+);
+
+/**
+ *
+ * @returns {boolean} whether or not the browser session supports sendBeacon
+ */
+const hasSendBeaconSupport = () => {
+  if (!navigator.sendBeacon || !document.visibilityState) {
+    return false;
+  }
+  return true;
 };
 
-function getStorage() {
+/**
+ * Fetch extra ad server data for a specific ad slot (bid)
+ * @param {object} bid an output from extractBid
+ * @returns {object} key value pairs from the adserver
+ */
+const getAdServerDataForBid = (bid) => {
+  const gptSlot = getGptSlotForAdUnitCode(bid);
+  if (gptSlot) {
+    return Object.fromEntries(
+      gptSlot
+        .getTargetingKeys()
+        .filter(
+          (key) =>
+            key.startsWith('pubx-') ||
+            (key.startsWith('hb_') && (key.match(/_/g) || []).length === 1)
+        )
+        .map((key) => [key, gptSlot.getTargeting(key)])
+    );
+  }
+  return {}; // TODO: support more ad servers
+};
+
+/**
+ * Access sessionStorage
+ * @returns {Storage}
+ */
+const getStorage = () => {
```

Comment:
This is a fairly serious rules violation, please correct immediately

Reply:
Please submit another pr fixing this issue or risk unplanned changes to your adapter, you may not bypass our stage methods

Reply:
We have updated this to use prebid's storage methods.